src/app/reducers/related-articles.handlers.ts

```diff
@@ -1,61 +1,48 @@
-import { cloneManuscript } from 'app/utils/state.utils';
 import { ManuscriptHistoryState } from 'app/store';
 import { RelatedArticle } from 'app/models/related-article';
-import { createDiff } from 'app/utils/history.utils';
+import { UpdateObjectChange } from 'app/utils/history/update-object-change';
+import { AddObjectChange } from 'app/utils/history/add-object-change';
+import { DeleteObjectChange } from 'app/utils/history/delete-object-change';
 
 export function updateRelatedArticle(state: ManuscriptHistoryState, payload: RelatedArticle): ManuscriptHistoryState {
   const articleIndex = state.data.present.relatedArticles.findIndex(({ id }) => id === payload.id);
 
-  const newDiff = createDiff({
-    relatedArticles: state.data.present.relatedArticles
-  });
-
-  const newManuscript = cloneManuscript(state.data.present);
-  newManuscript.relatedArticles[articleIndex] = payload;
-
+  const change = new UpdateObjectChange(
+    `relatedArticles.${articleIndex}`,
+    state.data.present.relatedArticles[articleIndex],
+    payload
+  );
   return {
     ...state,
     data: {
-      past: [...state.data.past, newDiff],
-      present: newManuscript,
+      past: [...state.data.past, change],
+      present: change.applyChange(state.data.present),
       future: []
     }
   };
 }
 
 export function addRelatedArticle(state: ManuscriptHistoryState, payload: RelatedArticle): ManuscriptHistoryState {
-  const newDiff = createDiff({
-    relatedArticles: state.data.present.relatedArticles
-  });
-
-  const newManuscript = cloneManuscript(state.data.present);
-  newManuscript.relatedArticles.push(payload);
+  const change = new AddObjectChange('relatedArticles', payload, 'id');
 
   return {
     ...state,
     data: {
-      past: [...state.data.past, newDiff],
-      present: newManuscript,
+      past: [...state.data.past, change],
+      present: change.applyChange(state.data.present),
       future: []
     }
   };
 }
 
 export function deleteRelatedArticle(state: ManuscriptHistoryState, payload: RelatedArticle): ManuscriptHistoryState {
-  const articleIndex = state.data.present.relatedArticles.findIndex(({ id }) => id === payload.id);
-
-  const newDiff = createDiff({
-    relatedArticles: state.data.present.relatedArticles
-  });
-
-  const newManuscript = cloneManuscript(state.data.present);
-  newManuscript.relatedArticles.splice(articleIndex, 1);
+  const change = new DeleteObjectChange('relatedArticles', payload, 'id');
```

Comment:
Same as keywords and references, add a new one, delete it, then use undo and it clears the list except for the one I just restored/added which is then duplicated.